src/3d/terrain/qgsdemterraintileloader_p.cpp

```diff
@@ -0,0 +1,235 @@
+
+  // get heightmap asynchronously
+  connect( generator->heightMapGenerator(), &QgsDemHeightMapGenerator::heightMapReady, this, &QgsDemTerrainTileLoader::onHeightMapReady );
+  mHeightMapJobId = generator->heightMapGenerator()->render( node->tileX(), node->tileY(), node->tileZ() );
+  mResolution = generator->heightMapGenerator()->resolution();
+}
+
+QgsDemTerrainTileLoader::~QgsDemTerrainTileLoader()
+{
+  qDebug() << "DEM chunk loader done.";
+}
+
+Qt3DCore::QEntity *QgsDemTerrainTileLoader::createEntity( Qt3DCore::QEntity *parent )
+{
+  QgsTerrainTileEntity *entity = new QgsTerrainTileEntity;
+
+  // create geometry renderer
+
+  Qt3DRender::QGeometryRenderer *mesh = new Qt3DRender::QGeometryRenderer;
+  mesh->setGeometry( new DemTerrainTileGeometry( mResolution, mHeightMap, mesh ) );
+  entity->addComponent( mesh ); // takes ownership if the component has no parent
+
+  // create material
+
+  createTextureComponent( entity );
+
+  // create transform
+
+  Qt3DCore::QTransform *transform;
+  transform = new Qt3DCore::QTransform();
+  entity->addComponent( transform );
+
+  float zMin, zMax;
+  _heightMapMinMax( mHeightMap, zMin, zMax );
+
+  const Qgs3DMapSettings &map = terrain()->map3D();
+  QgsRectangle extent = map.terrainGenerator()->tilingScheme().tileToExtent( mNode->tileX(), mNode->tileY(), mNode->tileZ() ); //node->extent;
+  double x0 = extent.xMinimum() - map.originX();
+  double y0 = extent.yMinimum() - map.originY();
+  double side = extent.width();
+  double half = side / 2;
+
+  transform->setScale3D( QVector3D( side, map.terrainVerticalScale(), side ) );
+  transform->setTranslation( QVector3D( x0 + half, 0, - ( y0 + half ) ) );
+
+  mNode->setExactBbox( QgsAABB( x0, zMin * map.terrainVerticalScale(), -y0, x0 + side, zMax * map.terrainVerticalScale(), -( y0 + side ) ) );
+
+  entity->setEnabled( false );
+  entity->setParent( parent );
+  return entity;
+}
+
+void QgsDemTerrainTileLoader::onHeightMapReady( int jobId, const QByteArray &heightMap )
+{
+  if ( mHeightMapJobId == jobId )
+  {
+    this->mHeightMap = heightMap;
+    mHeightMapJobId = -1;
+
+    // continue loading - texture
+    loadTexture();
+  }
+}
+
+
+// ---------------------
+
+#include <qgsrasterlayer.h>
+#include <QtConcurrent/QtConcurrentRun>
+#include <QFutureWatcher>
+
+QgsDemHeightMapGenerator::QgsDemHeightMapGenerator( QgsRasterLayer *dtm, const QgsTilingScheme &tilingScheme, int resolution )
+  : mDtm( dtm )
+  , mClonedProvider( ( QgsRasterDataProvider * )dtm->dataProvider()->clone() )
+  , mTilingScheme( tilingScheme )
+  , mResolution( resolution )
+  , mLastJobId( 0 )
+{
+}
+
+QgsDemHeightMapGenerator::~QgsDemHeightMapGenerator()
+{
+  delete mClonedProvider;
+}
+
+#include <QElapsedTimer>
+
+static QByteArray _readDtmData( QgsRasterDataProvider *provider, const QgsRectangle &extent, int res )
+{
+  QElapsedTimer t;
+  t.start();
+
+  // TODO: use feedback object? (but GDAL currently does not support cancelation anyway)
```

Comment:
may as well get this ready for when gdal provider does, then we'll get the benefit here for free...

Reply:
The problem is that we do not have a way to test the implementation with feedback object properly, so I would rather prefer to leave the TODO there...